SU2_CFD/include/numerics_simd/flow/convection/common.hpp

```diff
@@ -132,10 +132,12 @@ FORCEINLINE CPair<ReconVarType> reconstructPrimitives(Int iEdge, Int iPoint, Int
       break;
     }
     /*--- Detect a non-physical reconstruction based on negative pressure or density. ---*/
-    const Double neg_p_or_rho = fmax(fmin(V.i.pressure(), V.j.pressure()) < 0.0,
-                                     fmin(V.i.density(), V.j.density()) < 0.0);
+    /*--- Some weird issues with forward AD if this is all done in one line. ---*/
```

Comment:
What happened exactly?

Reply:
The result of the one line version was 1e-310, but I could not reproduce in a unit test, and also didnt get anything from valgrind. So i'm not sure, might have something to do with the expression types interfeering with each other?

Reply:
Instead of 0 or 1.

Reply:
I think I managed to reproduce it, I'll look into it a little further.

Reply:
Awesome thanks

Reply:
The issue is indeed buried deep inside expression templates. It boils down to a reference to a local variable.

We have
```
  VecExpr::max_::operator[] (size_t i) -> Scalar const& { return std::max(u[i], v[i]); }
```
with the reference-returning template `std::max`. Both `u[i]` and `v[i]` are returned by
```
  VecExpr::lt_::operator[] (size_t i) const -> Scalar { return Scalar(u[i] < v[i]); }
```
by value, so that `std::max` returns a reference to one of these local variables, which becomes dangling when returned by `max_::operator[]`.

Reply:
I'm using codi::max for the implementation of max_, is this falling through to std::max somehow?

Reply:
Yes. There is `using std::max` in the `codi` namespace, and the `std` template prevails against the overloads of `max` for CoDiPack expressions in this case.

Reply:
And is that also the case for reverse mode?

Reply:
The overall expression template structure is the same for forward and reverse, so I guess that it affects the reverse mode too.

Working towards a solution, using the version of `max` that returns a CoDiPack expression is not sufficient either, because the CoDiPack expression would capture the local `su2double` returned by
```
VecExpr::lt_::operator[] (size_t i) const -> Scalar { return Scalar(u[i] < v[i]); }
```
by an eventually dangling reference. Since everything is passive anyway,
```
VecExpr::lt_::operator[] (size_t i) const -> passivedouble { return passivedouble(u[i] < v[i]); }
```
is a better solution and CoDiPack expressions capture `passivedouble` correctly by value. However, this reintroduces the return-by-reference issue in the shape of `std::max` instantiated for `passivedouble`.

I think the best way out is using `codi::fmin` and `codi::fmax` together with the `passivedouble` change. As soon as `su2double` or other CoDiPack expressions are involved, they will always produce a CoDiPack expression, and the `cmath` versions of `fmin` and `fmax` return by value.

Reply:
I updated CoDiPack and made `std::fmin` and `std::fmax` available in `namespace codi` (next to overloads for them that were already there).

Mind if I add the fix suggested above?

Reply:
Fmin and fmax are not in std why was that needed? Im not sure I like that fix.

Reply:
Because fmin and fmax are not suitable implementations of min max for ints

Reply:
The `std` is not the point, just that writing 
```
VecExpr::max_::operator[] (size_t i) -> /*auto deduced*/ { return codi::fmax(u[i], v[i]); }
```
requires the `cmath` version of `fmax` to be accessible like this, too. [https://en.cppreference.com/w/cpp/numeric/math/fmax](https://en.cppreference.com/w/cpp/numeric/math/fmax) looks to me as if they are in `std`, though?

Reply:
True, the suggested solution with `passivedouble` and `fmin`/`fmax` is specific to the `su2double` instantiation of the vector expressions. But maybe we could introduce some small traits that choose what I suggested for `su2double` and default to the previous solution for everything else?

Reply:
Usually fmin and fmax work without namespace specification.
We can use fmin/max as the implementation for now and leave a note that they will not be efficient for integers.
I dont think we use that at the moment, Ill double check when I have some time and make the changes, thanks.

Reply:
Argument-dependent lookup should select the CoDiPack overloads automatically. Alright.

Reply:
I think the fix covers everything, active to passive, and suitable overloads.
Thank you for getting to the root cause.